mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -1740,3 +1741,294 @@ Value mlir::rock::addPassThroughIndices(OpBuilder &b, Value transformed,
+    std::pair<SmallVector<uint32_t>, SmallVector<uint32_t>> &removedDims) {
+  auto &sortedDeletedDims = std::get<Type>(removedDims);
+  llvm::sort(sortedDeletedDims);
+  for (auto &args : argsVector) {
+    auto &preserved = std::get<Type>(args.preservedDims);
+    for (auto [idx, dim] : llvm::enumerate(preserved)) {
+      size_t leastUpperBoundIdx = 0;
+      while ((leastUpperBoundIdx < sortedDeletedDims.size()) &&
+             (dim > sortedDeletedDims[leastUpperBoundIdx])) {
+        ++leastUpperBoundIdx;
+      }
+      preserved[idx] -= leastUpperBoundIdx;
+    }
+  }
+}
+
+FailureOr<rock::TransformMapAttr>
+removeUpperDimsFromMap(OpBuilder &b, rock::TransformMapAttr trMap,
+                       SetVector<int64_t> &removeIndicesSet,
+                       llvm::SmallVector<int64_t> &origUpperBounds,
+                       llvm::SmallVector<int64_t> &origLowerBounds) {
+
+  origLowerBounds =
+      llvm::SmallVector<int64_t>(trMap.getLowerBounds().asArrayRef());
+
+  SetVector<int64_t> newRemoveIndicesSet;
+  SmallVector<TransformAttr> newOps;
+  std::pair<SmallVector<uint32_t>, SmallVector<uint32_t>> removedDims = {};
+
+  std::vector<TransformAttrArgs> argsVector;
+  for (auto tr : trMap.getOps()) {
+    TransformAttrArgs args;
+    args.type = tr.getType();
+    SmallVector<uint32_t> &preservedUpperDims =
+        std::get<DimType::Upper>(args.preservedDims);
+    SmallVector<uint32_t> &preservedLowerDims =
+        std::get<DimType::Lower>(args.preservedDims);
+
+    preservedUpperDims =
+        getPreservedIndices<DimType::Upper>(tr, removeIndicesSet);
+
+    const bool mustBePreserved =
+        preservedUpperDims.size() == tr.getUpperDims().size();
+    const bool mustBeCompletelyRemoved = preservedUpperDims.empty();
+    const bool mustBeModified = !mustBePreserved && !mustBeCompletelyRemoved;
+
+    // compute which lower dimensions must be preserved
+    switch (args.type) {
+    case TransformType::AddDim:
+    case TransformType::ConstDim:
+    case TransformType::Broadcast:
+    case TransformType::Merge: {
+      assert(!mustBeModified && "assume always one-to-many relation");
+      [[fallthrough]];
+    }
+    case TransformType::Unmerge: {
+      if (mustBePreserved || mustBeModified) {
+        // preserve all original lower dims
+        llvm::copy(tr.getLowerDims(), std::back_inserter(preservedLowerDims));
+      }
+      break;
+    }
+    case TransformType::PassThrough: {
+      assert(tr.getUpperDims().size() == tr.getLowerDims().size());
+      auto localLowerDims = tr.getLowerDims();
+
+      // propagate all preserved upper dims to lower dims
+      for (auto [idx, dim] : llvm::enumerate(tr.getUpperDims())) {
+        if (llvm::is_contained(preservedUpperDims, dim)) {
+          preservedLowerDims.push_back(localLowerDims[idx]);
+        }
+      }
+      break;
+    }
+    default:
+      return failure();
+    }
+
+    propagatePreservedNames<DimType::Upper>(tr, args);
```

Comment:
I dont like this design pattern here (feel free to push back if you think it has merit).

I d rather make TransformAttrArgs struct filled correctly from construction.
i.e. I dont think we need to create it so early in L1831 and make it legal on the go.
Instead, it could be made to accept preservedDims as a mandatory arg for the constructor and use it fill preservedNames --> and remove propagatePreservedNames functions.

I d also make it const here to reduce error surface..